_posts/2021-08-02-Musing-on-Microservices.md

```diff
@@ -0,0 +1,29 @@
+---
+layout: post
+title: Musing on Microservices
+author: Charles Thomas
+---
+
+Microservices are commonplace among technology companies of all sizes and in this post, I hope to explore some of the pros and cons of them and what you should consider when thinking about whether or not they are a good solution for the problems you're solving.
+
+Let's start by quickly addressing a common question: what are microservices? Microservices are a way of structuring your application so that instead of having one big application (referred to as a monolith), you have multiple smaller applications that communicate with each other over a network.
+
+So why might you want to do this? Generally, microservices set out to solve two problems: scalability (how can I make my application handle more load?) and availability (how can I keep my application serving requests more of the time?). Let's deal with each of these issues separately.
+
+Microservices can help improve scalability by allowing you to run multiple copies of your app on different machines. However, you can do this with a monolith as well - just run more copies of the whole monolith  Instead, microservices provide a more subtle advantage: they allow you to scale different parts of your application separately according to demand. For example, you might have a part of your application such as authorization that has to process a very high number of requests while another part of your application such as settings has much lower traffic. In a microservices architecture, you can scale these parts of your application separately.
+
+In principle, this is great - it allows you to scale parts of your application independently, saving you money on hosting costs and allowing you to handle higher traffic loads more easily. However, the important question to ask yourself is whether or not different parts of your application actually have drastically different scalability needs and does the ability to scale them separately outweigh the cost of the added complexity?
+
+It is also worth asking what is currently limiting your ability to scale your application as a monolith? If it is CPU usage or memory, then perhaps, microservices may be a solution, however, you may be being bottlenecked by something else such as your database. In this case, microservices will not solve your problem, although the common pattern of having one database per microservice might help you. However, you can get the same effect by having your monolith connect to multiple databases.
```

Comment:
```suggestion
It is also worth asking what's currently limiting your ability to scale your monolith application? If it's CPU usage or memory, then perhaps, microservices may be a solution, however, you may be being bottlenecked by something else such as your database. In this case, microservices will not solve your problem, although the common pattern of having one database per microservice might help you. However, you can get the same effect by having your monolith connect to multiple databases.
```

Reply:
You also tend to [comma splice](https://examples.yourdictionary.com/comma-splice-examples.html#:~:text=A%20comma%20splice%20is%20a%20common%20grammatical%20error%20in%20English.&text=A%20comma%20splice%20is%20when,to%20identify%20an%20independent%20clause.) quite a bit, leading to long sentences that can be hard to parse